diff --git a/src/ddscxx/include/dds/topic/detail/TBuiltinTopicKeyImpl.hpp b/src/ddscxx/include/dds/topic/detail/TBuiltinTopicKeyImpl.hpp
index 0d05568..6147c1b 100644
--- a/src/ddscxx/include/dds/topic/detail/TBuiltinTopicKeyImpl.hpp
+++ b/src/ddscxx/include/dds/topic/detail/TBuiltinTopicKeyImpl.hpp
@@ -31,13 +31,13 @@ namespace topic
 template <typename D>
 const int32_t* TBuiltinTopicKey<D>::value() const
 {
-    return this->delegate().value();
+    return reinterpret_cast<const int32_t*>(this->delegate().value());
 }

 template <typename D>
 void TBuiltinTopicKey<D>::value(int32_t v[])
 {
-    return this->delegate().value(v);
+    this->delegate().value(reinterpret_cast<uint8_t*>(&v));
 }

 }
